### ISSUE TITLE
SHA=1 in-kernel hardware acceleration

### ISSUE THREAD

**Issue by neagix** [neagix/wii-linux-ngx#11]
Likewise #10, it should be possible to implemented SHA-1 in-kernel hardware acceleration.

For reference: http://wiibrew.org/wiki/Hardware/SHA-1_Engine
And: http://wiibrew.org/wiki/Hextwelve

Cc @marcan

**Comment by Jasjar42** [comment]
I don't think that will work. The Wii's SHA-1 engine isn't sane. It generates a completely different SHA-1 hash for a file than normal. It also sometimes generates a hash less than 40 bits long somehow.

**Comment by marcan** [comment]
Um, no, that's not the case. I have no idea where you've gotten that info from. The SHA-1 engine is completely compliant (though it does not perform padding, you have to do that yourself) and it always generates a correct 160-bit hash.

**Comment by neagix** [comment]
@Jasjar42 if you check the Hextwelve POC, it actually just runs a SHA1 unit test

**Comment by Jasjar42** [comment]
@marcan Sorry I was wrong. Can you tell me where my info is wrong? As wiibrew says boot0 loads boot1 and verifies its SHA1 hash against that stored in OTP. But the only boot1 OTP hash that is 40 bits long is the one for boot1d. Boot1b's OTP hash is: EF3EF781968D56DF5679A6F92E13F78BBDDFDF which is only 38 bits long. I wasn't able to match any SHA-1 hashes to the corresponding OTP hash in any way, even with encrypted or decrypted versions. Thanks to that I came to that conclusion.
